fem/fe/fe_pyramid.cpp

```diff
@@ -0,0 +1,1584 @@
+   dmu[ab-1] = -1.0 / (1.0 - z);
+   return dmu;
+}
+
+Vector FuentesPyramid::grad_mu1(real_t z, const Vector xy, unsigned int ab)
+{
+   Vector dmu({0.0, 0.0, xy[ab-1] / pow(1.0 - z, 2)});
+   dmu[ab-1] = 1.0 / (1.0 - z);
+   return dmu;
+}
+
+DenseMatrix FuentesPyramid::grad_mu01(real_t z, Vector xy, unsigned int ab)
+{
+   DenseMatrix dmu(2, 3);
+   dmu.SetRow(0, grad_mu0(z, xy, ab));
+   dmu.SetRow(1, grad_mu1(z, xy, ab));
+   return dmu;
+}
+
+Vector FuentesPyramid::mu01_grad_mu01(real_t z, Vector xy, unsigned int ab)
+{
+   Vector mu = mu01(z, xy, ab);
+   Vector mudmu(3);
+   add(mu(0), grad_mu1(z, xy, ab), -mu(1), grad_mu0(z, xy, ab), mudmu);
+   return mudmu;
+}
+
+Vector FuentesPyramid::grad_nu0(real_t z, const Vector xy, unsigned int ab)
+{
+   Vector dnu({0.0, 0.0, -1.0}); dnu[ab-1] = -1.0;
+   return dnu;
+}
+
+Vector FuentesPyramid::grad_nu1(real_t z, const Vector xy, unsigned int ab)
+{
+   Vector dnu({0.0, 0.0, 0.0}); dnu[ab-1] = 1.0;
+   return dnu;
+}
+
+Vector FuentesPyramid::grad_nu2(real_t z, const Vector xy, unsigned int ab)
+{
+   return Vector({0.0, 0.0, 1.0});
+}
+
+DenseMatrix FuentesPyramid::grad_nu01(real_t z, Vector xy, unsigned int ab)
+{
+   DenseMatrix dnu(2, 3);
+   dnu.SetRow(0, grad_nu0(z, xy, ab));
+   dnu.SetRow(1, grad_nu1(z, xy, ab));
+   return dnu;
+}
+
+DenseMatrix FuentesPyramid::grad_nu012(real_t z, Vector xy, unsigned int ab)
+{
+   DenseMatrix dnu(3, 3);
+   dnu.SetRow(0, grad_nu0(z, xy, ab));
+   dnu.SetRow(1, grad_nu1(z, xy, ab));
+   dnu.SetRow(2, grad_nu2(z, xy, ab));
+   return dnu;
+}
+
+DenseMatrix FuentesPyramid::grad_nu120(real_t z, Vector xy, unsigned int ab)
+{
+   DenseMatrix dnu(3, 3);
+   dnu.SetRow(0, grad_nu1(z, xy, ab));
+   dnu.SetRow(1, grad_nu2(z, xy, ab));
+   dnu.SetRow(2, grad_nu0(z, xy, ab));
+   return dnu;
+}
+
+Vector FuentesPyramid::nu01_grad_nu01(real_t z, Vector xy, unsigned int ab)
+{
+   Vector nu = nu01(z, xy, ab);
+   Vector nudnu(3);
+   add(nu(0), grad_nu1(z, xy, ab), -nu(1), grad_nu0(z, xy, ab), nudnu);
+   return nudnu;
+}
+
+Vector FuentesPyramid::nu12_grad_nu12(real_t z, Vector xy, unsigned int ab)
+{
+   Vector nu = nu12(z, xy, ab);
+   Vector nudnu(3);
+   add(nu(0), grad_nu2(z, xy, ab), -nu(1), grad_nu1(z, xy, ab), nudnu);
+   return nudnu;
+}
+
+Vector FuentesPyramid::nu012_grad_nu012(real_t z, Vector xy, unsigned int ab)
+{
+   Vector nu(nu012(z, xy, ab));
+   Vector dnu0(grad_nu0(z, xy, ab));
+   Vector dnu1(grad_nu1(z, xy, ab));
+   Vector dnu2(grad_nu2(z, xy, ab));
+
+   Vector v01(3), v12(3), v20(3);
+   dnu0.cross3D(dnu1, v01);
+   dnu1.cross3D(dnu2, v12);
+   dnu2.cross3D(dnu0, v20);
+
+   Vector nudnu(3);
+   add(nu(0), v12, nu(1), v20, nudnu);
+   nudnu.Add(nu(2), v01);
+   return nudnu;
+}
+
+void FuentesPyramid::CalcScaledLegendre(int p, real_t x, real_t t,
+                                        real_t *u)
+{
+   MFEM_ASSERT(p >= 0, "Polynomial order must be zero or larger");
+   if (t > 0.0)
+   {
+      Poly_1D::CalcLegendre(p, x / t, u);
```

Comment:
This could be numerically unstable when t is near 0

There is a recursive formula for these polynomials in the "Polynomials" table (Appendix E.1).